Java/core/src/main/java/com/amazon/randomcutforest/genericstore/Store.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.genericstore;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+
+/**
+ * This class provides a fixed amount of storage slots for a given type. When an
+ * object is stored, the class returns an index value that can be used to
+ * retrieve the object later. The purpose of this class is to reduce the overall
+ * memory footprint of an applicaiton by allowing classes to keep index values
+ * instead of pointers. The index values used in this class are ints, which
+ * occupy 4 bytes (Java pointers occupy 8 bytes). For further compression, see
+ * the {@link SmallStore} class which uses 2 byte short integers as index
+ * values.
+ *
+ * @param <T> The type of value being stored.
+ */
+public class Store<T> {
```

Comment:
question. should store depend on/reuse storecommon?

Reply:
This PR contains two independent implementations of the store concept: one using generics, and one using non-generic classes with primitive arrays. Eventually, we'll keep only one of the implementations (and in particular, we'll only have one or the other of Store<T> and StoreCommon).

Reply:
not an issue. the refactoring can be done later.